aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+
+For that reason, the recommendation is to use HTTPS for communications if your app is not in a trusted environment.
+
+## Cross-site scripting (XSS)
+This threat allows an unauthorized party to execute arbitrary logic in the context of the browser. A compromised application would potentially run arbitrary code on the client (user's browser) and such vulnerability could be used to potentially perform a number of actions against the server. Such actions include:
+* Dispatching fake/invalid events to the server.
+* Dispatching fail/invalid render completions.
+* Avoid dispatching render completions.
+* Dispatching interop calls from JS to .NET.
+* Modifying the response of interop calls from .NET to JS.
+* Avoid dispatching .NET to JS interop results.
+
+Blazor server-side takes some steps towards protecting against some of these threats but ultimately your application needs to be aware of these and take appropriate actions:
+* Don't trust data on event handlers before validating it first.
+* Take appropriate action upon receiving what your application considers to be invalid data.
+  * Ignore the data and return, if you can't be sure the input was illegitimate. This will allow the app to continue working.
+  * Throw an exception, if you determine that the input is illegitimate and couldn't have been produced by legitimate clients. This will tear down the circuit and end the session.
+* Don't trust the error message comming from render batch completions included in the logs. This error came from the client and can't be generally trusted as the client might have been compromised.
+* Don't trust the input from interop calls. This input comes in two flavors.
+  * JS to .NET interop invocations.
+  * .NET to JS interop results.
+* The application is responsible for validating that the contents of the arguments/result are valid even if they were correctly deserialized.
+
+Blazor server-side takes specific steps to mitigate some of these threats.
+* It will stop producing new UI updates if the client is not acknowledging render batches. (This can be configured in `CircuitOptions.MaxBufferedUnacknowledgedRenderBatches`)
+* It will timeout any .NET to JS call after 1 minute without receiving a response from the client.
+* It performs some basic validation in all input comming from the browser from JS interop.
+  * Ensure that the dotnet references are valid and of the type expected by the .NET method.
+  * Ensure that the data is not malformed.
+  * Ensure the right number of arguments for the method is present on the payload.
+  * Ensure that the arguments/result can be deserialized correctly before invoking the method.
+* It performs some basic validation in all input comming from the browser from dispatched events.
+  * It validates that the event has a valid type.
+  * It validates that the data for the event can be deserialized.
+  * It validates that there is an event handler associated with that event.
```

Comment:
Actually the entirety of this subsection up to this point doesn't seem to be about XSS. If this content isn't covered elsewhere, maybe it should go under a different heading.